unitary/quantum_chess/quantum_board.py

```diff
@@ -660,24 +665,72 @@ def _clear_path_ancilla(self, path_qubits, ancilla):
 
         This does the inverse operation of _create_path_ancilla.
         """
+        for q in path_qubits:
+            if q not in self.entangled_squares and nth_bit_of(
+                qubit_to_bit(q), self.state
+            ):
+                return
         self.circuit.append(qm.controlled_operation(cirq.X, [ancilla], [], path_qubits))
 
     def set_castle(self, sbit: int, rook_sbit: int, tbit: int, rook_tbit: int) -> None:
         """Adjusts classical bits for a castling operation."""
+        self.unhook(bit_to_qubit(tbit))
+        self.unhook(bit_to_qubit(rook_tbit))
         self.state = set_nth_bit(sbit, self.state, False)
         self.state = set_nth_bit(rook_sbit, self.state, False)
         self.state = set_nth_bit(tbit, self.state, True)
         self.state = set_nth_bit(rook_tbit, self.state, True)
 
-    def queenside_castle(
-        self, squbit: int, rook_squbit: int, tqubit: int, rook_tqubit: int, b_qubit: int
-    ) -> None:
-        """Performs a queenside castling operation."""
-        self.add_entangled(squbit, tqubit, rook_squbit, rook_tqubit)
-        self.circuit.append(
-            qm.queenside_castle(squbit, rook_squbit, tqubit, rook_tqubit, b_qubit)
+    def _do_noncontrolled_castle(
+        self, variant, measurement, sbit, tbit, rook_sbit, rook_tbit
+    ):
+        """Kingside castle, or queenside castle with empty b-file."""
+        squbit = bit_to_qubit(sbit)
+        tqubit = bit_to_qubit(tbit)
+        rook_squbit = bit_to_qubit(rook_sbit)
+        rook_tqubit = bit_to_qubit(rook_tbit)
+        f_empty = rook_tqubit not in self.entangled_squares and not nth_bit_of(
+            rook_tbit, self.state
         )
 
+        if variant == enums.MoveVariant.BASIC:
+            # Not in superposition, just castle
+            if f_empty:
+                self.set_castle(sbit, rook_sbit, tbit, rook_tbit)
+                return 1
+
+            self.state = set_nth_bit(sbit, self.state, False)
+            self.state = set_nth_bit(tbit, self.state, True)
+            self.add_entangled(rook_squbit, rook_tqubit)
+            self.circuit.append(qm.normal_move(rook_squbit, rook_tqubit))
+            return 1
+
+        elif variant == enums.MoveVariant.EXCLUDED:
```

Comment:
Maybe we could add some comments here explaining the two versions of the castle as you did in the PR?

Reply:
Done.